packages/deepsignal/core/src/index.ts

```diff
@@ -159,3 +109,155 @@ const shouldProxy = (val: any): boolean => {
 		(globalThis as any)[val.constructor.name] === val.constructor;
 	return !isBuiltIn || supported.has(val.constructor);
 };
+
+/***********/
+/** Types **/
+/***********/
+
+type DeepSignalObject<T extends object> = {
+	[P in keyof T & string as `$${P}`]?: Signal<T[P]>;
+} & {
+	[P in keyof T]: T[P] extends Array<unknown>
+		? DeepSignalArray<T[P]>
+		: T[P] extends object
+		? DeepSignalObject<T[P]>
+		: T[P];
+};
+
+// @ts-expect-error
```

Comment:
Do you have more context on why we need this? 

When I remove the `@ts-expect-error` I'm faced with this 
<img width="880" alt="Screenshot 2023-01-10 at 16 22 59" src="https://user-images.githubusercontent.com/5417266/211665291-28f028b8-a9f4-46cd-b257-61e8da8722b1.png">

Reply:
I guess it's like casting a type: "Hey TS, you think this is wrong but I know it's right, so shut up 🙂".

For some reason, this doesn't work:

```ts
type DeepArray<T> = Array<T> & { ... }
```

We could avoid extending `Array`:

```ts
interface DeepArray<T> { ... }
```

but then we'd have to copy all the methods that don't need to be modified. 

I also noticed that `ts-expect-error` is not copied when `tsc` creates the `.d.ts` file, but I don't know if `tsc` will fail on user projects or not, so I was thinking to merge this and find it out later 😅

Reply:
Related:

- https://github.com/microsoft/TypeScript/issues/43570
- https://github.com/microsoft/TypeScript/issues/38628

Reply:
Funny thing: if you use a `/** comment **` instead of `// comment` it is preserved in the `d.ts` file:

```ts
/** @ts-expect-error **/
interface DeepArray<T> extends Array<T> { ... }
```

I'll make that change and merge, and see what happens when people use this.